docs/source/developer_guide.md

```diff
@@ -8,11 +16,239 @@ The process for working on a CUDA/C++ feature which might span RAFT and one or m
+1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - see [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) as a reference for C++11  POD types.
+2. `raft::resources` - since it stores resource-related state which has nothing to do with model/algo state.
+3. Avoid using pointers to POD types (explicitly putting it out, even though it can be considered as a POD) and pass the structures by reference instead.
+   Internal to the C++ API, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.
+4. Accept single- (`raft::span`) and multi-dimensional views (`raft::mdspan`) and validate their metadata wherever possible.
+5. Prefer `std::optional` for any optional arguments (e.g. do not accept `nullptr`)
+6. All public APIs should be lightweight wrappers around calls to private APIs inside the `detail` namespace.
+
+### API stability
 
 Since RAFT is a core library with multiple consumers, it's important that the public APIs maintain stability across versions and any changes to them are done with caution, adding new functions and deprecating the old functions over a couple releases as necessary.
 
-The public APIs should be lightweight wrappers around calls to private APIs inside the `detail` namespace. 
+### Stateless C++ APIs
+
+Using the IVF-PQ algorithm as an example, the following way of exposing its API would be wrong according to the guidelines in this section, since it exposes a non-POD C++ class object in the C++ API:
+```cpp
+template <typename value_t, typename idx_t>
+class ivf_pq {
+  ivf_pq_params params_;
+  raft::resources const& res_;
+  
+public:
+  ivf_pq(raft::resources const& res);
+  void train(raft::device_matrix<value_t, idx_t, raft::row_major> dataset);
+  void search(raft::device_matrix<value_t, idx_t, raft::row_major> queries, 
+              raft::device_matrix<value_t, idx_t, raft::row_major> out_inds, 
+              raft::device_matrix<value_t, idx_t, raft::row_major> out_dists);
+};
+```
+
+An alternative correct way to expose this could be:
+```cpp
+namespace raft::ivf_pq {
+
+template<typename value_t, typename value_idx>
+void ivf_pq_train(raft::resources const& res, const raft::ivf_pq_params &params, raft::ivf_pq_index &index,
+raft::device_matrix<value_t, idx_t, raft::row_major> dataset);
+
+template<typename value_t, typename value_idx>
+void ivf_pq_search(raft::resources const& res, raft::ivf_pq_params const&params, raft::ivf_pq_index const & index,
+raft::device_matrix<value_t, idx_t, raft::row_major> queries,
+raft::device_matrix<value_t, idx_t, raft::row_major> out_inds,
+raft::device_matrix<value_t, idx_t, raft::row_major> out_dists);
+}
+```
+
+### Other functions on state
+
+These guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshalling) such a data structure. Further continuing the IVF-PQ example,  the following methods could achieve this:
```

Comment:
```suggestion
These guidelines also mean that it is the responsibility of C++ API to expose methods to load and store (aka marshaling) such a data structure. Further continuing the IVF-PQ example,  the following methods could achieve this:
```

Reply:
Ha, this looks like a conflict between UK english vs US english. I personally prefer `marshalling` as the verb form of `marshal`. It just feels right to me. However, it formally looks like another spelling is indeed `marshaling`.